x-pack/plugins/security_solution/public/onboarding/components/onboarding_body/cards/assistant/components/connectors/connector_cards.tsx

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import { type AIConnector } from '@kbn/elastic-assistant/impl/connectorland/connector_selector';
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiPanel,
+  EuiLoadingSpinner,
+  EuiText,
+  EuiBadge,
+  EuiSpacer,
+} from '@elastic/eui';
+import { css } from '@emotion/css';
+import { useKibana } from '../../../../../../../common/lib/kibana';
+import { CreateConnectorPopover } from './create_connector_popover';
+import { ConnectorSetup } from './connector_setup';
+
+interface ConnectorCardsProps {
+  connectors?: AIConnector[];
+  onConnectorSaved: () => void;
+  onClose?: () => void;
+}
+
+export const ConnectorCards = React.memo<ConnectorCardsProps>(
+  ({ connectors, onConnectorSaved, onClose }) => {
+    const {
+      triggersActionsUi: { actionTypeRegistry },
+    } = useKibana().services;
+
+    if (!connectors) return <EuiLoadingSpinner />;
+
+    if (connectors.length > 0) {
+      return (
+        <>
+          <EuiFlexGroup
+            wrap
+            className={css`
+              max-height: 290px;
+              overflow-y: auto;
+            `}
+          >
+            {connectors?.map((connector) => (
```

Comment:
nit:
- doesn't need the `?.` here since the type is array here
- should be wrapped in a `useMemo`. Since it's not possible here as it's after conditional I'd argue it's better to move the early return part with spinner out into the parent or make it as part of a HoC that this component is wrapped in before export.

Reply:
> doesn't need the ?. here since the type is array here

Yup 👍 

> should be wrapped in a useMemo. Since it's not possible here as it's after conditional I'd argue it's better to move the early return part with spinner out into the parent or make it as part of a HoC that this component is wrapped in before export.

Do we really need to wrap it? I'll add the missing `key` prop, that should be enough.

Reply:
There's no need to memoize components in a `map`, as long as they have the `key`. Since we assume every component is memorized, there's no need to do double memoization.

Reply:
it's not a double memoize, they serve different levels, your map will still run, your react.createElement will still run, if later you add more code into closure it will still happen. I agree that dom is the expensive part and its improved. But it's not the whole thing that's memoized. That's why I mentioned different levels of memoization. And frankly I don't  care about microoptimization unless it's benchmarked and proven my only point is if we have a convention and I raise it that's it. Feel free to resolve.

Reply:
I am not saying that using the `key` and the `useMemo` would mean double memoization. What I am saying is that all components are `React.memo`'ed (by convention), so nothing would happen if React tries to update them with the same props. And that's only true when the `key` is correctly set.

The only thing the `useMemo` would do here is prevent the array loop when the `connectors` prop reference has not changed and the component receives an update (by some other prop), but even in this case would have any impact on the underlying component rendering, since is already taken care of. That's a marginal improvement.

Reply:
1. I hope you understand that your words 
![image](https://github.com/user-attachments/assets/bef65bca-8a76-4e3e-a936-1e9fe9e95b26)
say one thing and you say you didn't say that is breeding confusion. Maybe you didn't mean what you said but you said it and I respond to what you say while trying to interpret what you mean in good faith.

2. You are defending a point I never argued with you 
![image](https://github.com/user-attachments/assets/0f848b2b-35c6-4e91-b905-8cf12d60a225)
I never said it will affect underlying rendering. If you read my messages carefully you will notice that I am saying `useMemo` is preventing .map from happening, which prevents react.createElement from happening. And it's a marginal singular improvement here which I agree with you from the start. But that was never the point. The point was that if there is a **convention**, convention **includes** marginal or even miniscule improvements as well, especially when there is **no harm or downside** in using useMemo. And on scale when you look at the entire application all these tiny marginal things add up and **make a difference**, that's why community and us probably have adopted the convention. And I not sitting in this debate to argue about the value I am simply pattern matching from convention similarly to how you did with context in my PR and many other things. This is our job as reviewers. But when I do that you oppose for whatever reason and enter into polemics about efficiency which is better settled in engineering sync with multiple perspectives, don't you think?

3. Also notice how similar your argumentation for not using `useMemo` is to my argumentation in my PR for using useContext with useMemo provider object, namely the fact that everything around is memoized, so there is marginal impact. But when you do it as a reviewer I have to just consider because it's a convention, yet when I do the same to you in your PR suddenly it's debatable. That doesn't seem fair and it's confusing for me personally to write the new code and to get it reviewed next time by you.